### ISSUE TITLE
Bitbucket Server Authentication Failed

### ISSUE THREAD

**Issue by trahloff** [terraform-aws-modules/terraform-aws-atlantis#96]
I basically have the same problem that was discussed in #54 but with Bitbucket Server, instead of Bitbucket Cloud. 

Since the original issue was closed without resolution and no further input, I would like to investigate this issue more in-depth since I was able to reproduce it. 

I set these params:
~~~
  atlantis_bitbucket_user       = "<user>"
  atlantis_bitbucket_user_token = "<password>"
~~~

In addition, I introduced the Terraform variable "atlantis_bitbucket_base_url" to set the environment parameter "ATLANTIS_BITBUCKET_BASE_URL" which is needed to run Bitbucket Server with Atlantis (see https://github.com/runatlantis/atlantis/issues/622).

I used this custom workflow:

```yaml
# atlantis.yaml
version: 3
projects:
- dir: .
  workflow: echo
workflows:
  echo:
    plan:
      steps:
      - run: echo $ATLANTIS_BITBUCKET_USER
      - run: echo $ATLANTIS_BITBUCKET_TOKEN
```

_Originally posted by @lkysow in https://github.com/terraform-aws-modules/terraform-aws-atlantis/issues/54#issuecomment-499515606_

I verified access via cURL: 

~~~
curl --location --request GET 
'https://<BitbucketServerEndpoint>/rest/api/1.0/projects/<BitbucketProject>/repos/infrastructure/pull-requests/23/changes?start=0' \
--header 'Authorization: Bearer <Token>'
~~~

(works)

This is the output after opening a new PR:

~~~json
 [DBUG] events_controller.go:122 server: Handling Bitbucket Server post
 [DBUG] events_controller.go:208 server: Handling as pull request state changed event
 [INFO] events_controller.go:304 server: Identified event as type "opened"
 [INFO] events_controller.go:345 server: Executing autoplan
 [INFO] middleware.go:42 server: POST /events – respond HTTP 200
 [WARN] command_runner.go:114 <BitbucketProject+Repo>#23: Unable to update commit status: making request "POST https://<BitbucketServer>/rest/build-status/1.0/commits/081a96ff9e6d001c56db8eb70d824008c54ad8d8" unexpected status code: 401, body: 
        {
            "errors": [
                {
                    "context": null,
                    "message": "Authentication failed. Please check your credentials and try again.",
                    "exceptionName": "com.atlassian.bitbucket.auth.IncorrectPasswordAuthenticationException"
                }
            ]
        }
 [DBUG] project_command_builder.go:106 <BitbucketProject+Repo>#23: Got workspace lock
 [WARN] command_runner.go:120 <BitbucketProject+Repo>#23: Unable to update commit status: making request "POST https://<BitbucketServer>/rest/build-status/1.0/commits/081a96ff9e6d001c56db8eb70d824008c54ad8d8" unexpected status code: 401, body: 
        {
            "errors": [
                {
                    "context": null,
                    "message": "Authentication failed. Please check your credentials and try again.",
                    "exceptionName": "com.atlassian.bitbucket.auth.IncorrectPasswordAuthenticationException"
                }
            ]
        }
 [EROR] command_runner.go:419 <BitbucketProject+Repo>#23: Making request "GET https://<BitbucketServer>/rest/api/1.0/projects/<BitbucketProject>/repos/infrastructure/pull-requests/23/changes?start=0" unexpected status code: 401, body: 
        {
            "errors": [
                {
                    "context": null,
                    "message": "Authentication failed. Please check your credentials and try again.",
                    "exceptionName": "com.atlassian.bitbucket.auth.IncorrectPasswordAuthenticationException"
                }
            ]
        }
 [EROR] command_runner.go:426 <BitbucketProject+Repo>#23: Unable to comment: making request "POST https://<BitbucketServer>/rest/api/1.0/projects/<BitbucketProject>/repos/infrastructure/pull-requests/23/comments" unexpected status code: 401, body: 
        {
            "errors": [
                {
                    "context": null,
                    "message": "Authentication failed. Please check your credentials and try again.",
                    "exceptionName": "com.atlassian.bitbucket.auth.IncorrectPasswordAuthenticationException"
                }
            ]
        }
~~~

**Comment by trahloff** [comment]
I also tried to manually add a comment via cURL which worked without a problem: 

~~~bash
curl -D- -X POST -H "Authorization: Bearer <token>" -H "Content-Type: application/json" https://<BitbucketServer>/rest/api/1.0/projects/<BitbucketProject>/repos/test/pull-requests/1/comments -d '{
         "text": "An insightful general comment on a pull request."
     }'
~~~

**Comment by lkysow** [comment]
@trahloff Atlantis isn't using a bearer token, it's use HTTP basic auth: https://github.com/runatlantis/atlantis/blob/master/server/events/vcs/bitbucketserver/client.go#L268-L281 have you tried your curl using that?

**Comment by trahloff** [comment]
@lkysow HTTP basic auth works perfectly fine via curl. I can only find documentation on how to use Bitbucket [Personal Access Tokens](https://www.runatlantis.io/docs/access-credentials.html#generating-an-access-token) with Atlantis and I can only find documentation for [config flags](https://www.runatlantis.io/docs/server-configuration.html#flags) that set Bitbucket user and *token*. Any idea how I could switch to user/password authentication? Did I miss something?

**Comment by lkysow** [comment]
@trahloff is it possible that there are spurious characters added to your username or password that's being passed to Atlantis? Maybe an extra newline or space or something? You could use a custom workflow like https://github.com/terraform-aws-modules/terraform-aws-atlantis/issues/54#issuecomment-499515606 to verify.

**Comment by trahloff** [comment]
@lkysow I added the custom workflow but since the authentication is not working Atlantis is not able to get the PR changes and thus cannot get the commit files and exits before it can execute the workflow.

I understood from your comment that:
`ATLANTIS_BITBUCKET_TOKEN` != Bitbucket Personal Access Token and
`ATLANTIS_BITBUCKET_TOKEN` == Bitbucket Basic Auth Password

Is that correct?

**Comment by asmontas-i** [comment]
I think I had exact same problem just recently. The problem was bad credentials as the error message suggested. 

@trahloff, are you able to get a shell of your running container and do a request via curl using **the same** env variables (`ATLANTIS_BITBUCKET_USER`  and `ATLANTIS_BITBUCKET_TOKEN`) that are set in your container? 

Btw: `ATLANTIS_BITBUCKET_TOKEN` == Bitbucket Personal Access Token 
and I believe you can use it as a Bearer token as per [BBS docs here](https://confluence.atlassian.com/bitbucketserver/personal-access-tokens-939515499.html). and as a basic auth password.

**Comment by lkysow** [comment]
No sorry, I was just saying that if you wanted to reproduce the calls in the same way as Atlantis you should be using HTTP basic auth, not the bearer token header.

I've created a custom build of Atlantis that echoes out the credentials upon startup: https://github.com/runatlantis/atlantis/commit/2f6e90ee31cb7930818f5f36bfd96b13edbaa681

[atlantis_linux_arm.zip](https://github.com/terraform-aws-modules/terraform-aws-atlantis/files/4220663/atlantis_linux_arm.zip)
[atlantis_linux_amd64.zip](https://github.com/terraform-aws-modules/terraform-aws-atlantis/files/4220664/atlantis_linux_amd64.zip)
[atlantis_linux_386.zip](https://github.com/terraform-aws-modules/terraform-aws-atlantis/files/4220665/atlantis_linux_386.zip)
[atlantis_darwin_amd64.zip](https://github.com/terraform-aws-modules/terraform-aws-atlantis/files/4220666/atlantis_darwin_amd64.zip)

Do you want to try that?

**Comment by trahloff** [comment]
Thank you guys for your support! 

@lkysow The custom firmware worked like a charm. The credentials were correct. 

I tried to set `ATLANTIS_BITBUCKET_TOKEN` to my Bitbucket *password* and that works. Weird that the Personal Access Token did not. 

Unfortunately, I ran into the "SCM access over HTTP(S) has been disabled" error but that may be an indicator that our BBS instance has some kind of restrictive configuration that interfered with the Personal Access Token solution. I guess the problem is on our side and is not caused by Atlantis or the Terraform module. 

Sorry guys and thank you again for helping me.

**Comment by lkysow** [comment]
No worries, glad that you've gotten a bit further. I don't have easy access to bitbucket server to test myself at the moment. If it turns out that for some configurations the personal access token doesn't work then a doc update PR would be awesome.

**Comment by trahloff** [comment]
I will definitely investigate further and make a doc update PR if I find something. 

@asmontas-i Are you interested in a Terraform PR adding the `atlantis_bitbucket_base_url` for BBS configuration or would you like to handle this configuration via the `custom_environment_variables` variable?

**Comment by asmontas-i** [comment]
@trahloff, I see this already have been [taken care of](https://github.com/terraform-aws-modules/terraform-aws-atlantis/commit/dc905daa17d7396d95e27a7834dc69d395d8a356) :))

**Comment by antonbabenko** [comment]
As @asmontas-i saying, this has been resolved in #79 and is available in [v2.5.0](https://github.com/terraform-aws-modules/terraform-aws-atlantis/releases/v2.5.0) of this module.

I am closing this issue unless I am missing something.
